iml-agent/src/action_plugins/ntp/action_configure.rs

```diff
@@ -0,0 +1,376 @@
+// Copyright (c) 2019 DDN. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+use crate::agent_error::ImlAgentError;
+use futures::{future, Future, Stream, TryFutureExt, TryStreamExt};
+use std::path::Path;
+
+static NTP_CONFIG_FILE: &str = "/etc/ntp.conf";
+static MARKER: &str = "# IML_EDIT";
+static REMOVE_MARKER: &str = "#REMOVE_MARKER#";
+static PREFIX: &str = "server";
+
+/// Gets a stream to the ntp config
+fn get_ntp_config_stream() -> impl Stream<Item = Result<String, ImlAgentError>> {
+    iml_fs::stream_file_lines(Path::new(NTP_CONFIG_FILE)).err_into()
+}
+
+/// Writes the new config data to the config file
+pub async fn update_and_write_new_config(server: Option<String>) -> Result<(), ImlAgentError> {
+    let s = get_ntp_config_stream();
+
+    let updated_config = configure_ntp(server, s).await?;
+
+    tokio::fs::write(NTP_CONFIG_FILE, updated_config).await?;
+
+    Ok(())
+}
+
+fn configure_ntp(
+    server: Option<String>,
+    s: impl Stream<Item = Result<String, ImlAgentError>>,
+) -> impl Future<Output = Result<String, ImlAgentError>> {
+    s.map_ok(reset_config)
+        .try_filter(|x| future::ready(filter_out_remove_markers(x)))
+        .try_fold(
+            ("".to_string(), false),
+            move |(mut acc, prefix_found), line| {
+                let (x, prefix_found) = if let Some(server_name) = &server {
+                    transform_config(server_name.into(), line, prefix_found)
+                } else {
+                    (line, prefix_found)
+                };
+
+                acc.push_str(&format!("{}\n", x));
```

Comment:
I think `push_str(x)` followed by `push_char('\n')` should be cheaper. Avoid allocations.

Reply:
Updated.